core/src/main/java/org/elasticsearch/index/query/WrapperQueryParser.java

```diff
@@ -41,28 +39,33 @@ public WrapperQueryParser() {
     }
 
     @Override
-    public Query parse(QueryParseContext parseContext) throws IOException, QueryParsingException {
+    public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOException, QueryParsingException {
         XContentParser parser = parseContext.parser();
 
-        XContentParser.Token token = parser.nextToken();
-        if (token != XContentParser.Token.FIELD_NAME) {
-            throw new QueryParsingException(parseContext, "[wrapper] query malformed");
-        }
-        String fieldName = parser.currentName();
-        if (!fieldName.equals("query")) {
-            throw new QueryParsingException(parseContext, "[wrapper] query malformed");
+        byte[] source = null;
+        float boost = AbstractQueryBuilder.DEFAULT_BOOST;
+        String queryName = null;
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if ("query".equals(currentFieldName)) {
+                    source = parser.binaryValue();
+                } else {
+                    throw new QueryParsingException(parseContext, "[wrapper] query does not support [" + currentFieldName + "]");
+                }
```

Comment:
yea  the `parser.binaryValue()` that you introduced instead of parsing the inner query is fine. But I don't think we need a loop if we parse a single field, that's what I meant when I say to back to what we did before.